llama_index/query_engine/jsonalyze_query_engine.py

```diff
@@ -0,0 +1,236 @@
+        service_context (Optional[ServiceContext]): The service context.
+        table_name (str): The table name to use, defaults to DEFAULT_TABLE_NAME.
+
+    Returns:
+        Tuple[str, Dict[str, Any], List[Dict[str, Any]]]: The SQL Query,
+            the Schema, and the Result.
+    """
+    try:
+        import sqlite_utils
+    except ImportError as exc:
+        IMPORT_ERROR_MSG = (
+            "sqlite-utils is needed to use this Query Engine:\n"
+            "pip install sqlite-utils"
+        )
+
+        raise ImportError(IMPORT_ERROR_MSG) from exc
+    # Instantiate in-memory SQLite database
+    db = sqlite_utils.Database(memory=True)
+    try:
+        # Load list of dictionaries into SQLite database
+        db[table_name].insert_all(list_of_dict)
+    except sqlite_utils.db_exceptions.IntegrityError as exc:
+        print_text(f"Error inserting into table {table_name}, expected format:")
+        print_text("[{col1: val1, col2: val2, ...}, ...]")
+        raise ValueError("Invalid list_of_dict") from exc
+
+    # Get the table schema
+    table_schema = db[table_name].columns_dict
+
+    # Get the SQL query with text-to-SQL prompt
+    sql_query = service_context.llm.predict(
+        prompt,
+        table_name=table_name,
+        table_schema=table_schema,
+        question=query,
+    )
+
+    try:
+        # Execute the SQL query
+        results = list(db.query(sql_query))
+    except sqlite_utils.db_exceptions.OperationalError as exc:
+        print_text(f"Error executing query: {sql_query}")
+        raise ValueError("Invalid query") from exc
+
+    return sql_query, table_schema, results
+
+
+class JSONalyzeQueryEngine(BaseQueryEngine):
+    """JSON List Shape Data Analysis Query Engine.
+
+    Converts natural language statasical queries to SQL within in-mem SQLite queries.
+
+    list_of_dict(List[Dict[str, Any]]): List of dictionaries to query.
+    service_context (ServiceContext): ServiceContext
+    jsonalyze_prompt (BasePromptTemplate): The JSONalyze prompt to use.
+    analyzer (Callable): The analyzer that executes the query.
+    synthesize_response (bool): Whether to synthesize a response.
+    response_synthesis_prompt (BasePromptTemplate): The response synthesis prompt
+        to use.
+    table_name (str): The table name to use.
+    verbose (bool): Whether to print verbose output.
+    """
+
+    def __init__(
+        self,
+        list_of_dict: List[Dict[str, Any]],
+        service_context: ServiceContext,
+        jsonalyze_prompt: Optional[BasePromptTemplate] = None,
+        analyzer: Optional[Callable] = None,
+        synthesize_response: bool = True,
+        response_synthesis_prompt: Optional[BasePromptTemplate] = None,
+        table_name: str = DEFAULT_TABLE_NAME,
+        verbose: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize params."""
+        self._list_of_dict = list_of_dict
+        self._service_context = service_context
```

Comment:
nit: should be `self._service_context = service_context or ServiceContext.from_defaults()`

Reply:
Thanks, sorry for that, done!